Support contain-intrinsic-size auto && <length>

cathiechen opened the pull request:
#### 88587e7bd5451f3504cfc7576d2f7617ab1b7cf8
<pre>
Support contain-intrinsic-size auto &amp;&amp; &lt;length&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=239450">https://bugs.webkit.org/show_bug.cgi?id=239450</a>

Reviewed by Oriol Brufau.

This patch adds support to value auto &amp;&amp; &lt;length&gt; for contain-intrinsic-size. Per [1], with the auto &amp;&amp; &lt;length&gt; value,
the last remembered size is the explicit intrinsic size if the box is skipping its content.
The last remembered size is determined when ResizeObserver events are delivered. m_resizeObserverForContainIntrinsicSize
in Document is created specifically to handle the last remembered size. CallbackForContainIntrinsicSize in Document is
used to handle the lastRememberedSize. Because this ResizeObserver is for inside use, so we create a interface
NativeResizeObserverCallback for native use. Now in ResizeObserver there are two kinds of callbacks:
JSResizeObserverCallback and NativeResizeObserverCallback, only one could be initialized.
The resizeObserver would observe the target if contain-intrinsic-size is valued auto &amp;&amp; &lt;length&gt;, otherwise unobserve
it and clear the lastRememberedSize. If an observed target is hidden or skipping its content, unobserve it. Similarly,
if an observed target is removed from DOM tree, unobserve it and clear the lastRememberedSize.

[1] <a href="https://www.w3.org/TR/css-sizing-4/#valdef-contain-intrinsic-width-auto--length">https://www.w3.org/TR/css-sizing-4/#valdef-contain-intrinsic-width-auto--length</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-011-expected.txt:
* LayoutTests/resize-observer/contain-instrinsic-size-should-not-leak-nodes-expected.txt: Added.
* LayoutTests/resize-observer/contain-instrinsic-size-should-not-leak-nodes.html: Added.
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
(WebCore::Document::gatherResizeObservationsForContainIntrinsicSize):
(WebCore::Document::deliverResizeObservations):
(WebCore::Document::updateResizeObservations):
(WebCore::Document::ensureResizeObserverForContainIntrinsicSize):
(WebCore::Document::observeForContainIntrinsicSize):
(WebCore::Document::unobserveForContainIntrinsicSize):
(WebCore::Document::resetObservationSizeForContainIntrinsicSize):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/ResizeObservation.cpp:
(WebCore::ResizeObservation::resetObservationSize):
* Source/WebCore/page/ResizeObservation.h:
* Source/WebCore/page/ResizeObserver.cpp:
(WebCore::ResizeObserver::create):
(WebCore::ResizeObserver::createNativeObserver):
(WebCore::ResizeObserver::ResizeObserver):
(WebCore::ResizeObserver::observeInternal):
(WebCore::ResizeObserver::observe):
(WebCore::ResizeObserver::deliverObservations):
(WebCore::ResizeObserver::isJSCallback):
(WebCore::ResizeObserver::isNativeCallback):
(WebCore::ResizeObserver::callbackConcurrently):
(WebCore::ResizeObserver::resetObservationSize):
* Source/WebCore/page/ResizeObserver.h:
(WebCore::ResizeObserver::callbackConcurrently): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::hasAutoLengthContainIntrinsicSize const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: <a href="https://commits.webkit.org/260535@main">https://commits.webkit.org/260535@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eada6ab89059c81db2c8f34d92895b1d95c4a3b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108161 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41028 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117285 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116603 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112047 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8529 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100368 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113928 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14073 "Found 6 new test failures: editing/deleting/smart-delete-003.html, editing/pasteboard/smart-paste-003-trailing-whitespace.html, editing/pasteboard/smart-paste-005.html, editing/pasteboard/smart-paste-paragraph-004.html, editing/selection/doubleclick-whitespace-img-crash.html, editing/selection/ios/selection-handles-in-iframe.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41957 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95983 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28883 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83628 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10109 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30231 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10832 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7133 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16258 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49822 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7281 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12414 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -196,4 +222,23 @@ bool ResizeObserver::removeObservation(const Element& target)
+                continue;
+
+            auto contentBoxeSize = entry->contentBoxSize().at(0);
+            target->saveLastRememberedSize(ResizeObserverSize::create(contentBoxeSize->inlineSize(), contentBoxeSize->blockSize()));
```
Maybe only one of `contain-intrinsic-width` and `contain-intrinsic-height` has the `auto` value. The last remembered size should be tracked independently for each axis (block and inline), see https://github.com/w3c/csswg-drafts/issues/7529. That's why test `auto-009.html` is still failing.

Can be done in a follow-up, just FYI.

cathiechen replied:
@Loirooriol Thanks a lot!
Yes, let me fix it in the follow-up patch:)

heycam commented on Source/WebCore/page/ResizeObserver.h:
```diff
@@ -51,12 +51,21 @@ struct ResizeObserverData {
     bool hasActiveObservations() const { return m_activeObservations.size(); }
 
+    void observeForContainIntrinsicSize(Element&);
+    void observeInternal(Element&, const ResizeObserverBoxOptions);
```
`observeInternal` should be private.

And since `ResizeObserverBoxOptions` is just an enum, I don't think we need to mark it `const`.

cathiechen replied:
Done, made it private.

The `const` is passed from `void ResizeObserver::observe(Element& target, const ResizeObserverOptions& options)`.
Should I remove both?

heycam commented on Source/WebCore/page/ResizeObserver.h:
```diff
@@ -73,12 +82,17 @@ class ResizeObserver : public RefCounted<ResizeObserver>, public CanMakeWeakPtr<
     WeakPtr<Document, WeakPtrImplWithEventTargetData> m_document;
+    // Either m_callbackForContainIntrinsicSize or m_callback should be non-empty.
+    RefPtr<CallbackForContainIntrinsicSize> m_callbackForContainIntrinsicSize;
     RefPtr<ResizeObserverCallback> m_callback;
```
Can we make this a std::variant so that we can only hold one or the other?

cathiechen replied:
Done!

heycam commented on Source/WebCore/page/ResizeObserver.h:
```diff
@@ -51,12 +51,21 @@ struct ResizeObserverData {
+        void onResize(const Vector<Ref<ResizeObserverEntry>>&);
+    };
+
+    static Ref<ResizeObserver> create(Document&);
```
Is it possible to have any contain-intrinsic-size specific behavior live somewhere else? While contain-intrinsic-size internally relies on ResizeObserver to 

* making this class called something more generic, like `NativeResizeObserverCallback`
* making its `onResize` function virtual
* defining `CallbackForContainIntrinsicSize` in Document.cpp, inheriting from `NativeResizeObserverCallback`, where we do the registration / deregistration of the observer
* passing a `Ref<NativeResizeObserverCallback>&&` into the `ResizeObserver` constructor

I think if we can avoid tangling in contain-intrinsic-size specific behavior in the ResizeObserver implementation, that will help keep things a bit clearer.

cathiechen replied:
Done! Thanks!

heycam commented on Source/WebCore/page/ResizeObservation.cpp:
```diff
@@ -118,6 +119,19 @@ std::optional<ResizeObservation::BoxSizes> ResizeObservation::elementSizeChanged
+            if (LayoutSize(lastRememberedSize->inlineSize(), lastRememberedSize->blockSize()) != currentSizes->contentBoxLogicalSize)
+                return currentSizes;
+            return { };
+        }
```
I don't understand why we need to have this contain-intrinsic-size specific behavior here. Why can't we rely on the normal "compare last observation size to current size" check, and then have the caller (`CallbackForContainIntrinsicSize::onResize`) update the last remembered size if needed?

I think it would be better if we could avoid having c-i-s specific logic inside `ResizeObservation` if possible.

cathiechen replied:
Yes, this is necessary. There are some differences between them. For instance,
1. `lastRememberedSize` is not always same to `m_lastObservationSizes`, we skip updating `lastRememberedSize` when it is size containment.
2. RO would not trigger delivery if the size is same to the initial size. But c-i-s should, for it need to create `lastRememberedSize`.
3. When target is removed from DOM tree, c-i-s needs to clear  `lastRememberedSize` and disconnect, but RO not.

cathiechen replied:
Maybe I should find a way to specify the size comparing function?

Loirooriol replied:
That's how I did these in Firefox:

1. Just observe or unobserve the element when the condition for updating the last remembered size changes
2. If you mean that no observation is triggered when observing a 0x0 element, the CSSWG resolved to change that in general, so c-i-s shouldn't need a special behavior
3. I handled this in `Element::UnbindFromTree` (probably WebKit has something similar): unobserving and observing the element schedules a notification, even if the size was already 0x0. Then the callback checks if the element is disconnected, if so it removes the last remembered size.

cathiechen replied:
Thanks, @Loirooriol ! Let me have a try of this!
Looks like we need to address the initial size issue in RO first.
Just uploaded the patch https://github.com/WebKit/WebKit/pull/8839

cathiechen replied:
Sorry for not updating recently, but I'm stuck in this issue https://github.com/WebKit/WebKit/pull/9112#issuecomment-1403885997

cathiechen replied:
OK, 2 is now fixed:)
As to 1 and 3, instead of unobserving the target, we update the m_lastObservationSizes in the new patch.
For 1, set m_lastObservationSizes to lastRememberedSize.
For 3, set update m_lastObservationSizes to -1x-1, to ensure that the disconnected target could be delivered.

heycam commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -196,4 +222,23 @@ bool ResizeObserver::removeObservation(const Element& target)
+            if (auto* box = target->renderBox(); box && box->shouldApplySizeContainment())
+                continue;
+
+            auto contentBoxeSize = entry->contentBoxSize().at(0);
```
`contentBoxSize`

cathiechen replied:
Done

heycam commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -196,4 +222,23 @@ bool ResizeObserver::removeObservation(const Element& target)
+                continue;
+
+            auto contentBoxeSize = entry->contentBoxSize().at(0);
+            target->setLastRememberedSize(ResizeObserverSize::create(contentBoxeSize->inlineSize(), contentBoxeSize->blockSize()));
```
I know this was in the previous patch, but looking at this now, do we need to store this refcounted `ResizeObserverSize` object in the `ElementRareData`? Maybe a `std::optional<FloatSize>` is OK?

cathiechen replied:
Yes, I think `std::optional` would work. And we may need `std::optional<Float>` here, for there could be only c-i-s in one direction. The WPT auto-009.html tests this scenario. I have a patch to fix it, we could merge the `std::optional` change to it. Both find to me. Do you want me to use `std::optional` in this patch or in the other patch?
https://bugs.webkit.org/show_bug.cgi?id=249919

cathiechen reviewed (commented):
@heycam Thanks a lot for the review!
I addressed most of the comments. The one in ResizeObservation is not, I will keep working on it.

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,33 @@ struct Document::PendingScrollEventTargetList {
+            if (auto* box = target->renderBox(); !box || box->shouldApplySizeContainment()) {
+                auto* lastRememberedSize = target->lastRememberedSize();
+                LayoutSize size = lastRememberedSize ? LayoutSize(lastRememberedSize->inlineSize(), lastRememberedSize->blockSize()) : LayoutSize(-1, -1);
+                target->document().updateObservationSizeForContainIntrinsicSize(*target, size);
```
I don't understand the logic here. If the element doesn't have a last remembered size, then this sets the observation to -1x-1 to schedule another notification, which will likely trigger the same logic and keep running this code every frame even if nothing changes? Why?
And  if the element has a last remembered size from before getting size containment, this sets the observation size to the last remembered size, which is probably different, so it also schedules notifications.

cathiechen replied:
Yes, that's true, it will be triggered every frame if there is no change.
From this point of view, look like the "unobserve" solution is move efficient.
OK, I will work on it:)

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,33 @@ struct Document::PendingScrollEventTargetList {
+                continue;
+            }
+
+            if (auto* box = target->renderBox(); !box || box->shouldApplySizeContainment()) {
```
BTW, you should check for "skips contents" instead of "has size containment". See the resolution in https://github.com/w3c/csswg-drafts/issues/7807#issuecomment-1317388716

Loirooriol commented on Source/WebCore/page/ResizeObservation.cpp:
```diff
@@ -53,6 +53,18 @@ void ResizeObservation::updateObservationSize(const BoxSizes& boxSizes)
+        m_lastObservationSizes.borderBoxLogicalSize = LayoutSize(-1, -1);
+        return;
+    case ResizeObserverBoxOptions::ContentBox:
+        m_lastObservationSizes.contentBoxLogicalSize = LayoutSize(-1, -1);
```
For `contain-intrinsic-size` clearing `contentBoxLogicalSize` seems enough, but in general it seems weird to only clear either `contentBoxLogicalSize` or `borderBoxLogicalSize` (and never `contentBoxSize`). Maybe just do

```cpp
m_lastObservationSizes = { LayoutSize(-1, -1), LayoutSize(-1, -1), LayoutSize(-1, -1) };
```

cathiechen replied:
OK, done, thanks!

Loirooriol commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -677,6 +677,16 @@ ExceptionOr<void> ContainerNode::removeChild(Node& oldChild)
+            document().observeForContainIntrinsicSize(element);
+            document().resetObservationSizeForContainIntrinsicSize(element);
+        }
+    }
```
You can use `dynamicDowncast`:

```suggestion
    auto* element = dynamicDowncast<Element>(oldChild);
    if (element && element->lastRememberedSize()) {
        // The disconnected element could be unobserved because of other properties, here we need to make sure it is observed,
        // so that deliver could be triggered and it would clear lastRememberedSize.
        document().observeForContainIntrinsicSize(*element);
        document().resetObservationSizeForContainIntrinsicSize(*element);
    }
```

cathiechen replied:
Done, thanks!

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -45,9 +45,28 @@ Ref<ResizeObserver> ResizeObserver::create(Document& document, Ref<ResizeObserve
+    : m_document(document)
+    , m_JSOrNativeCallback(JSOrNativeCallback(nullptr, WTFMove(nativeCallback)))
 {
 }
```
I think this could be

```cpp
Ref<ResizeObserver> ResizeObserver::create(Document& document, Ref<ResizeObserverCallback>&& callback)
{
    return adoptRef(*new ResizeObserver(document, callback.ptr()));
}

Ref<ResizeObserver> ResizeObserver::createNativeObserver(Document& document, Ref<NativeResizeObserverCallback>&& nativeCallback)
{
    return adoptRef(*new ResizeObserver(document, nativeCallback.ptr()));
}

ResizeObserver::ResizeObserver(Document& document, JSOrNativeResizeObserverCallback&& callback)
    : m_document(document)
    , m_JSOrNativeCallback(WTFMove(callback))
{
}
```

cathiechen replied:
Done, thanks! Looks much better!

cathiechen replied:
BTW, I use `adoptRef(*new ResizeObserver(document,  { RefPtr<ResizeObserverCallback> { WTFMove(callback) } }));` to avoid `callback`'s ref count churn.

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -148,17 +174,22 @@ void ResizeObserver::deliverObservations()
 
     InspectorInstrumentation::willFireObserverCallback(*context, "ResizeObserver"_s);
-    m_callback->handleEvent(*this, entries, *this);
+    std::get<RefPtr<ResizeObserverCallback>>(m_JSOrNativeCallback)->handleEvent(*this, entries, *this);
```
This should probably keep asserting `hasCallback()`

```cpp
    // FIXME: The JSResizeObserver wrapper should be kept alive as long as the resize observer can fire events.
    ASSERT(isJSCallback());
    auto jsCallback = std::get<RefPtr<ResizeObserverCallback>>(m_JSOrNativeCallback);
    ASSERT(jsCallback->hasCallback());
    if (!jsCallback->hasCallback())
        return;

    auto* context = jsCallback->scriptExecutionContext();
    if (!context)
        return;

    InspectorInstrumentation::willFireObserverCallback(*context, "ResizeObserver"_s);
    jsCallback->handleEvent(*this, entries, *this);
```

cathiechen replied:
Done, thanks!

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -58,20 +77,16 @@ ResizeObserver::~ResizeObserver()
-void ResizeObserver::observe(Element& target, const ResizeObserverOptions& options)
+void ResizeObserver::observeInternal(Element& target, const ResizeObserverBoxOptions boxOptions)
 {
-    if (!m_callback)
-        return;
```
Why remove this?

cathiechen replied:
Done, added `ASSERT(m_JSOrNativeCallback.valueless_by_exception() == false);`

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,33 @@ struct Document::PendingScrollEventTargetList {
+                target->document().unobserveForContainIntrinsicSize(*target);
+                continue;
+            }
+
```
Maybe assert `!box->shouldSkipContent()`.

cathiechen replied:
Done

Loirooriol replied:
Can also be good to assert `box->style().hasAutoLengthContainIntrinsicSize()`

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,33 @@ struct Document::PendingScrollEventTargetList {
+            }
+
+            auto contentBoxSize = entry->contentBoxSize().at(0);
+            target->setLastRememberedSize(ResizeObserverSize::create(contentBoxSize->inlineSize(), contentBoxSize->blockSize()));
```
The test for `display: inline` not clearing the last remembered size seems to pass, but where are you handling that? Seems like this line would set it to 0x0.

cathiechen replied:
`target->renderBox()` is null for `display: inline`, so it `unobserveForContainIntrinsicSize`, but not clear the lastRememberedSize, and won't update the value.

cathiechen reviewed (commented):
@Loirooriol Thanks a lot for the review!

Loirooriol commented on Source/WebCore/page/ResizeObserver.h:
```diff
@@ -49,14 +49,25 @@ struct ResizeObserverData {
+    NativeResizeObserverCallback() = default;
+    virtual ~NativeResizeObserverCallback() = default;
+    virtual void onResize(const Vector<Ref<ResizeObserverEntry>>&) = 0;
+};
```
I think you can remove this class, use

```cpp
using NativeResizeObserverCallback = void (*)(const Vector<Ref<ResizeObserverEntry>>&);
using JSOrNativeResizeObserverCallback = std::variant<RefPtr<ResizeObserverCallback>, NativeResizeObserverCallback>;
```

And in Document.cpp

```cpp
void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry>>& entries)
{
    //  ...
}
ResizeObserver::createNativeObserver(*this, CallbackForContainIntrinsicSize);
```

cathiechen replied:
Done, added `ResizeObserver*`, so that in the callback we could call `observer->unobserve(*target)` instead:
```
using NativeResizeObserverCallback = void (*)(const Vector<Ref<ResizeObserverEntry>>&, ResizeObserver*);
using JSOrNativeResizeObserverCallback = std::variant<RefPtr<ResizeObserverCallback>, NativeResizeObserverCallback>;
```

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,34 @@ struct Document::PendingScrollEventTargetList {
+    for (auto& entry : entries) {
+        if (auto* target = entry->target()) {
+            if (!target->isConnected()) {
+                target->document().unobserveForContainIntrinsicSize(*target);
```
Nit: ResizeObserver callbacks typically receive the observer as an argument. If you pass it too for native callbacks, here you can just use `observer.unobserve(*target);` without needing to call `document()` nor `unobserveForContainIntrinsicSize()`.

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8490,14 +8531,17 @@ void Document::setHasSkippedResizeObservations(bool skipped)
     // and it could change other frame in deliverResizeObservations().
     page.layoutIfNeeded();
 
     // Start check resize obervers;
```
While you are at it, this line had a typo (observers)

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -42,12 +42,17 @@ namespace WebCore {
 Ref<ResizeObserver> ResizeObserver::create(Document& document, Ref<ResizeObserverCallback>&& callback)
 {
-    return adoptRef(*new ResizeObserver(document, WTFMove(callback)));
+    return adoptRef(*new ResizeObserver(document,  { RefPtr<ResizeObserverCallback> { WTFMove(callback) } }));
```
Typo: you used 2 spaces after the comma. And shouldn't this work too?

```cpp
return adoptRef(*new ResizeObserver(document, { WTFMove(callback) }));
```

cathiechen replied:
Done, and nope, the compiler complains.

cathiechen reviewed (commented):
Hi @Loirooriol , thanks a lot for the review!
Let's see if other reviewers have questions about this patch, if not, I will land this patch tomorrow:)

Loirooriol commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -383,6 +384,31 @@ struct Document::PendingScrollEventTargetList {
 
+static void CallbackForContainIntrinsicSize(const Vector<Ref<ResizeObserverEntry>>& entries, ResizeObserver* observer)
+{
+    ASSERT(observer);
```
You can pass `ResizeObserver&` which will make it clearer that it's not null.

Loirooriol commented on Source/WebCore/page/ResizeObserver.cpp:
```diff
@@ -148,17 +162,24 @@ void ResizeObserver::deliverObservations()
     auto targetsWaitingForFirstObservation = std::exchange(m_targetsWaitingForFirstObservation, { });
 
+    if (isNativeCallback()) {
+        std::get<NativeResizeObserverCallback>(m_JSOrNativeCallback)(entries, this);
```
`(entries, *this)` would be closer to the JS case

cathiechen replied:
done

cathiechen committed:
Support contain-intrinsic-size auto && <length>
https://bugs.webkit.org/show_bug.cgi?id=239450

Reviewed by Oriol Brufau.

This patch adds support to value auto && <length> for contain-intrinsic-size. Per [1], with the auto && <length> value,
the last remembered size is the explicit intrinsic size if the box is skipping its content.
The last remembered size is determined when ResizeObserver events are delivered. m_resizeObserverForContainIntrinsicSize
in Document is created specifically to handle the last remembered size. CallbackForContainIntrinsicSize in Document is
used to handle the lastRememberedSize. Because this ResizeObserver is for inside use, so we create a interface
NativeResizeObserverCallback for native use. Now in ResizeObserver there are two kinds of callbacks:
JSResizeObserverCallback and NativeResizeObserverCallback, only one could be initialized.
The resizeObserver would observe the target if contain-intrinsic-size is valued auto && <length>, otherwise unobserve
it and clear the lastRememberedSize. If an observed target is hidden or skipping its content, unobserve it. Similarly,
if an observed target is removed from DOM tree, unobserve it and clear the lastRememberedSize.

[1] https://www.w3.org/TR/css-sizing-4/#valdef-contain-intrinsic-width-auto--length

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-011-expected.txt:
* LayoutTests/resize-observer/contain-instrinsic-size-should-not-leak-nodes-expected.txt: Added.
* LayoutTests/resize-observer/contain-instrinsic-size-should-not-leak-nodes.html: Added.
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/Document.cpp:
(WebCore::CallbackForContainIntrinsicSize):
(WebCore::Document::gatherResizeObservationsForContainIntrinsicSize):
(WebCore::Document::deliverResizeObservations):
(WebCore::Document::updateResizeObservations):
(WebCore::Document::ensureResizeObserverForContainIntrinsicSize):
(WebCore::Document::observeForContainIntrinsicSize):
(WebCore::Document::unobserveForContainIntrinsicSize):
(WebCore::Document::resetObservationSizeForContainIntrinsicSize):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/ResizeObservation.cpp:
(WebCore::ResizeObservation::resetObservationSize):
* Source/WebCore/page/ResizeObservation.h:
* Source/WebCore/page/ResizeObserver.cpp:
(WebCore::ResizeObserver::create):
(WebCore::ResizeObserver::createNativeObserver):
(WebCore::ResizeObserver::ResizeObserver):
(WebCore::ResizeObserver::observeInternal):
(WebCore::ResizeObserver::observe):
(WebCore::ResizeObserver::deliverObservations):
(WebCore::ResizeObserver::isJSCallback):
(WebCore::ResizeObserver::isNativeCallback):
(WebCore::ResizeObserver::callbackConcurrently):
(WebCore::ResizeObserver::resetObservationSize):
* Source/WebCore/page/ResizeObserver.h:
(WebCore::ResizeObserver::callbackConcurrently): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::hasAutoLengthContainIntrinsicSize const):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: https://commits.webkit.org/260535@main

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -677,6 +677,14 @@ ExceptionOr<void> ContainerNode::removeChild(Node& oldChild)
     rebuildSVGExtensionsElementsIfNecessary();
     dispatchSubtreeModifiedEvent();
 
+    auto* element = dynamicDowncast<Element>(oldChild);
```
This is a wrong place to add this kind of logic. We need to have this code in `ContainerNode::removeAllChildrenWithScriptAssertion` and `ContainerNode::removeNodeWithScriptAssertion`, not this one function.

cathiechen replied:
Created a pr for this. https://github.com/WebKit/WebKit/pull/25247
